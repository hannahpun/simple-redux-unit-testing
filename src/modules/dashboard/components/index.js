import { DashboardContainer as DashboardComponent } from './containers/mappings';

export { DashboardComponent };
