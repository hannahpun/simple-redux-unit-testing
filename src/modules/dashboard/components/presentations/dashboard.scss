$red: #c0392b;
$black: #2d3436;
$f-base: 18px;


.form-group{
    display: flex;
    flex-wrap: wrap;
    font-size: $f-base;
    width: 260px;
    label{
        width: 100%;
        display: block;
        text-align: left;
        font-weight: bold;
        color: #636e72;
        background-color: #fff;
        margin-bottom: 10px;
    }
    .form-control{
        width: 100%;
        border: 3px solid;
        padding: 5px 10px;
        font-size: $f-base*1.4;
    }
    .form-text{
        margin-top: 5px;
        display: none;
    }

    &.invalid{
        color: $red;
        .form-control{
            border-color: $red;
        }
        .form-text{
            display: block;
        }
    }
}