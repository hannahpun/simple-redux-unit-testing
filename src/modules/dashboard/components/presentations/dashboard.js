import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './dashboard.scss';
import { render } from '@testing-library/react';

class DashboardView extends Component {
    state = {
        username: '',
        hasSubmit: false
    };

    handleChange = (e) => {
        this.setState({ username: e.target.value, hasSubmit: false });
    };
    handleKeyDown = (e) => {
        if (e.key === 'Enter') {
            const { auth } = this.props;
            auth(this.state.username);
            this.setState({ hasSubmit: true });
        }
    };
    render () {
        const { login } = this.props;

        return (
            <div data-test='dashboard' className='dashboard'>
                {login ? (
                    <h1 data-test='welcome-text'>Howdy, {this.state.username} !</h1>
                ) : (
                    <div data-test='form-group' className={`form-group ${this.state.hasSubmit ? 'invalid' : ''}`}>
                        <label htmlFor='username'>Please Enter Your Username</label>
                        <input
                            data-test='input'
                            className='form-control'
                            id='username'
                            type='text'
                            value={this.state.username}
                            onChange={this.handleChange}
                            onKeyDown={this.handleKeyDown}
                        />
                        <small data-test='form-text' className='form-text text-muted'>
                            Wrong Username
                        </small>
                    </div>
                )}
            </div>
        );
    }
}

DashboardView.propTypes = {
    login: PropTypes.bool.isRequired,
    auth: PropTypes.func
};

export { DashboardView };
