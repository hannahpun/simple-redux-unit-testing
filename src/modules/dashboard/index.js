import * as actions from './actions';
import * as components from './components';
import * as reducer from './reducers';

export { actions, components, reducer };
