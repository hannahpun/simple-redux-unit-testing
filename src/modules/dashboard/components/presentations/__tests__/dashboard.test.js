import React from 'react';
import { shallow } from 'enzyme';
import { findByTestAttr } from 'test/testUtils';

import { DashboardView } from '../dashboard';

const defaultProps = { login: false };
/**
 * Factory function to create a ShallowWrapper for the Congrats component.
 * @function setup
 * @param {object} props - Component props specific to this setup.
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
    const setupProps = { ...defaultProps, ...props };

    return shallow(<DashboardView {...setupProps} />);
};

describe('Render correct component before login', () => {
    let wrapper = null;

    beforeEach(() => {
        wrapper = setup();
        wrapper.setState({ hasSubmit: false });
    });

    test('render without error', () => {
        const appComponent = findByTestAttr(wrapper, 'dashboard');
        expect(appComponent.exists()).toBe(true);
    });
    test('render <input> ', () => {
        const appComponent = findByTestAttr(wrapper, 'form-group');
        expect(appComponent.length).toBe(1);
    });
    test('does not render welcome text ', () => {
        const appComponent = findByTestAttr(wrapper, 'welcome-text');
        expect(appComponent.exists()).toBe(false);
    });
});

describe('Render correct component after login', () => {
    test('render welcome text', () => {
        const wrapper = setup({ login: true });
        const appComponent = findByTestAttr(wrapper, 'welcome-text');
        expect(appComponent.exists()).toBe(true);
    });
});

describe('Handler function', () => {
    let wrapper = null;
    let instance = null;

    beforeEach(() => {
        wrapper = setup({ auth: jest.fn() });
        wrapper.setState({ hasSubmit: false });
        instance = wrapper.instance();
    });
    test('simulate onchange - handleChange', () => {
        const appComponent = findByTestAttr(wrapper, 'input');
        const expectedEvent = {
            target: { value: 'testing state' }
        };

        appComponent.simulate('change', expectedEvent);
        expect(wrapper.state('username')).toBe('testing state');
    });

    describe('function - handleKeyDown', () => {
        test('when key equal enter ', () => {
            const expectedEvent = {
                key: 'Enter'
            };
            instance.handleKeyDown(expectedEvent);
            expect(wrapper.state('hasSubmit')).toBe(true);
        });
        test('when key does not equal enter ', () => {
            const expectedEvent = {
                key: 'Space'
            };
            instance.handleKeyDown(expectedEvent);
            expect(wrapper.state('hasSubmit')).toBe(false);
        });
    });
});

describe('show correct class name in form-group', () => {
    test('Show invalid class name when hasSubmit is true', () => {
        const wrapper = setup();
        wrapper.setState({ hasSubmit: true });
        const appComponent = findByTestAttr(wrapper, 'form-group').hasClass('invalid');
        expect(appComponent).toBe(true);
    });
});
