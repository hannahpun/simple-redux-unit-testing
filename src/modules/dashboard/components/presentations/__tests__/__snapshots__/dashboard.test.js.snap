// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`After login render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <h1
    data-test="welcome-text"
  >
    Howdy, 
     !
  </h1>
</div>
`;

exports[`Before login render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <div
    className="form-group "
    data-test="form-group"
  >
    <label
      htmlFor="username"
    >
      Please Enter Your Username
    </label>
    <input
      className="form-control"
      id="username"
      type="text"
    />
    <small
      className="form-text text-muted"
      data-test="form-text"
    >
      Wrong Username
    </small>
  </div>
</div>
`;

exports[`before login render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <div
    className="form-group "
  >
    <label
      htmlFor="username"
    >
      Please Enter Your Username
    </label>
    <input
      className="form-control"
      data-test="input"
      id="username"
      type="text"
    />
    <small
      className="form-text text-muted"
      data-test="form-text"
    >
      Wrong Username
    </small>
  </div>
</div>
`;

exports[`render before login render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <div
    className="form-group "
  >
    <label
      htmlFor="username"
    >
      Please Enter Your Username
    </label>
    <input
      className="form-control"
      data-test="input"
      id="username"
      type="text"
    />
    <small
      className="form-text text-muted"
      data-test="form-text"
    >
      Wrong Username
    </small>
  </div>
</div>
`;

exports[`render render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <div
    className="form-group "
  >
    <label
      htmlFor="username"
    >
      Please Enter Your Username
    </label>
    <input
      className="form-control"
      data-test="input"
      id="username"
      type="text"
    />
    <small
      className="form-text text-muted"
      data-test="form-text"
    >
      Wrong Username
    </small>
  </div>
</div>
`;

exports[`render without error 1`] = `
<div
  className="dashboard"
  data-test="dashboard"
>
  <div
    className="form-group "
  >
    <label
      htmlFor="username"
    >
      Please Enter Your Username
    </label>
    <input
      className="form-control"
      data-test="input"
      id="username"
      type="text"
    />
    <small
      className="form-text text-muted"
      data-test="form-text"
    >
      Wrong Username
    </small>
  </div>
</div>
`;
